### ISSUE TITLE
Enhancement: E7 bank switch expansion

### ISSUE THREAD

**Issue by mohhingman** [stella-emu/stella#759]
An enhancement idea for submission.

The E7 bankswitch scheme is very nice, especially for layering ram and rom within the same memory space simultaneously. I was wondering if it's possible to expand the scheme to offer 32KB of ROM and 4KB of RAM in Stella.  The e7 design is great, though expansion could benefit significantly for larger games using E7 architecture to have additional music, levels and graphics. The RAM/ROM limits above were chosen inline with the capability of the Harmony Cartridge. 

~~~~~~~~~~~~~~~~~~~~~
Currently E7 trigger points are:

$FFE0 ;Bank 0 (2K ROM)
$FFE1 ;Bank 1 (2K ROM)
$FFE2 ;Bank 2 (2K ROM)
$FFE3 ;Bank 3 (2K ROM)
$FFE4 ;Bank 4 (2K ROM)
$FFE5 ;Bank 5 (2K ROM)
$FFE6 ;Bank 6 (2K ROM)
$FFE7 ;1K RAM Bank (in 2K space for read/write)
$FFE8 ;256b RAM 0 (in 512b space for read/write)
$FFE9 ;256b RAM 1 (in 512b space for read/write)
$FFEA ;256b RAM 2 (in 512b space for read/write)
$FFEB ;256b RAM 3 (in 512b space for read/write)

~~~~~~~~~~~~~~~~~~

Perhaps the 16 addresses before the current trigger ($FFD0-$FFDF) points could trigger additional banks and ram, whilst the original addresses remain the same? Suggestion below.

$FFD0 ;Bank 7 (2K ROM)
$FFD1 ;Bank 8 (2K ROM)
$FFD2 ;Bank 9 (2K ROM)
$FFD3 ;Bank A (2K ROM)
$FFD4 ;Bank B (2K ROM)
$FFD5 ;Bank C (2K ROM)
$FFD6 ;Bank D (2K ROM)
$FFD7 ;Bank E (2K ROM)
$FFD8 ;256b RAM 4 (in 512b space for read/write)
$FFD9 ;256b RAM 5 (in 512b space for read/write)
$FFDA ;256b RAM 6 (in 512b space for read/write)
$FFDB ;256b RAM 7 (in 512b space for read/write)

The last 4 RAM banks 4-7 could alternatively become another 1K Ram bank. by letting $FFD8 trigger that. My preference is for 4 x 256b banks extra instead of a 1k bank.  

Thank you for your consideration.

**Comment by thrust26** [comment]
Interesting, but does any hardware which supports this scheme exist?

**Comment by mohhingman** [comment]
>

> Interesting, but does any hardware which supports this scheme exist?

No. This is only a development idea at this point.

**Comment by thrust26** [comment]
I suppose we have to put this at the end of our to-do until there is hard- or software.

**Comment by mohhingman** [comment]
OK. I will prepare a test ROM and source code.

**Comment by mohhingman** [comment]
Update for D7 bank scheme. Have been able to install Stella source and build locally. Developing "CartD7.cxx" and "CartD7.hxx", and possibly augmenting "CartMNetwork.cxx/hxx" (yet TBD). Am planning to change the original proposal slightly as below, to be easier to implement the banking in the Cart code and the trigger points have been made more contiguous. Also forgot about the other 1K RAM on last post, so that will become a 1K bank. New proposal below. It could take a while to fully test and develop, no need to hold up next release from this idea.

D7 Banking Scheme (15x 2k ROM banks, 2 x 1k RAM banks, 8 x 256b RAM banks, 1.5kB fixed bank at end)

$FFD7 ;1K RAM Bank (in 2K space for read/write) (NEW)
$FFD8 ;Bank 0 (2K ROM)
$FFD9 ;Bank 1 (2K ROM)
$FFDA ;Bank 2 (2K ROM)
$FFDB ;Bank 3 (2K ROM)
$FFDC ;Bank 4 (2K ROM)
$FFDD ;Bank 5 (2K ROM)
$FFDE ;Bank 6 (2K ROM)
$FFDF ;Bank 7 (2K ROM)
$FFE0 ;Bank 8 (2K ROM)
$FFE1 ;Bank 9 (2K ROM)
$FFE2 ;Bank A (2K ROM)
$FFE3 ;Bank B (2K ROM)
$FFE4 ;Bank C (2K ROM)
$FFE5 ;Bank D (2K ROM)
$FFE6 ;Bank E (2K ROM)
$FFE7 ;1K RAM Bank (in 2K space for read/write)
$FFE8 ;256b RAM 0 (in 512b space for read/write)
$FFE9 ;256b RAM 1 (in 512b space for read/write)
$FFEA ;256b RAM 2 (in 512b space for read/write)
$FFEB ;256b RAM 3 (in 512b space for read/write)
$FFDC ;256b RAM 4 (in 512b space for read/write)
$FFDD ;256b RAM 5 (in 512b space for read/write)
$FFDE ;256b RAM 6 (in 512b space for read/write)
$FFDF ;256b RAM 7 (in 512b space for read/write)

**Comment by mohhingman** [comment]
update on this. Have finished basic Stella code for the new scheme/cart detector/debugger and a test rom. There are some thoughts of improving the last 1.5kb space for adding functionality for patching in read only 256 ram pages. Will update again when this is complete.

**Comment by mohhingman** [comment]
An update on progress. 

It is apparent the code base for the M Network Carts was written to support different ROM sized carts flexibly, such that derived classes (such as E7 and E78K) can customize the image size and structure based on their ROM size with the RAM structure appended on the end of the image. However, the code in the M Network cart and GUI Widget classes for RAM banking are hard coded to support a single 1K bank and the 4x 256b banks. To make the number of banks flexible, the development action so far has been to upgrade the CartMNetwork and CartMNetworkWidget classes to support further arguments in their Initialize function to pass customized the quantities of the 2 types of RAM banks. This is in particular for the VariantLists. Also, overriding the internalramsize constant. So far, have tested a new 32k variant with 2 x 1K banks, and 8 x 256b banks which appears to work as expected in emulation and in the debugger. Have also tested previous E7 and E78k ROMs that work OK. 

The ultimate plan is to create a more advanced E7 based scheme that apart from increased rom and ram capabilities could also have new segments in the last 1.5kb for small 256b read only pages that reference the 256b ram pages. These segments would be primarily for  absolute addressed graphics data for kernels to reference. Over successive frames the banks could be changed to create animation, where the kernel still picks out the same address but now different graphics data. This is to allow the vblank/overscan to minimize time working with graphics data and more time for processing logic. Additionally, the small segments could allow code changes in the last bank during vblank/overscan which traditionally has been the bottlenecked by fixed rom code. The implication of the slots is the need for new GUI variantlist handlers in Stella. Finally, the support of the scheme with hardware is uncertain, especially on mature products such as Encore. It may be that this scheme would exist as an emulation experiment for some time.

The planned outputs of these developments are as follows:

E732k - A simple derived class that is identical to E78k and E7 that allows 32k of ROM and the RAM is unchanged, for maximizing the ROM space for perhaps Harmony with the minimum number of changes required to work.  The application is to allow E7 games to double the ROM capacity with very little changes.
E8 - Similar to E732k, but can support 2 x 1K RAM banks and 8 x 256b RAM banks, with as little changes as possible, This would theoretically maximize the banking scheme in the harmony cart. This would allow E7 games more ram and rom space.
EX - For supporting large ROM sizes - possibly 384k, and large RAM sizes (possibly 32k) with the feature of patching in the read only 256b segments in the last bank. The application of such a scheme would be for large tile based games with significant graphics, script and music data (such as an RPG). 

The question that may be asked is that, is this large scheme necessary? Can another scheme work? The closest is 3E+ and is already supported by hardware. This scheme is almost ideal and is a step in the right direction. The main lacking component is the limit of 64kb ROM. The 3E scheme is in terms of ROM size ideal, but lacks the flexibility of RAM to RAM / RAM to ROM read/write/execution that E7 and 3E+ have. There is a slight edge that E7 possesses in the speed of ROM and RAM swapouts for self modifying code, where it can change a ROM/RAM bank in 3 cycles as opposed to 5 cycles for a 3E scheme. E7 can also use the $FFEx trigger read to load a constant stored there that can be pertinent in the next subroutine that changed banks (such as a counter).  

I will continue to develop the 256b segments idea for the EX scheme and report back on progress. The customization of the 1k/256b RAM pages is complete for now.

**Comment by mohhingman** [comment]
Some further thoughts on the E7 expansion ideas. The scheme of using the 256b patched ram banks in the last 1.5kb is a large departure from E7 so should should be segregated from E7 schemes. On closer examination of the 3E/3E+ schemes, it would appear more efficient to make a hybrid of 3E and E7 techniques to maximize ROM/RAM size and save on wasted toggle addresses. 

Therefore, I'm planning to split the ideas into two developments -1. E7 ROM/RAM expansions (no segmentation changes) derived from CartMNetwork class, and  2. The Above 3E/E7 hybrid idea, which would probably be derived from Cart3E class instead and with some borrowed code from CartMNetwork.

**Comment by mohhingman** [comment]
An evaluation of 3E+'s suitability has taken place. It is possible to assign one of the 4 banks that shifts between kernel and main loop  and then have the other 3 banks are used as independent sprite pattern/colour banks. This is also supported on the Pluscart platform and verification of a 64kb ROM size running has been achieved on Pluscart. For these reasons, the 3E/E7 hybrid idea is shelved for the time being in favour of using 3E+.  

As for the  E7 expansion ideas, I'm planning to cut this back to just a 32k version of E7 (with no ram changes).

**Comment by mohhingman** [comment]
Hello Admins,

This ticket can be cleared.

I have decided to move away from this plan to expand E7 bankswitching. It seemed like a good idea at first, however I think it's best to leave it as is.

Some reasons for this:
- It would be best not to upset the legacy games detection heuristics
- Based on what i've experimented with the Stella code, it required a large, non-trivial re-write of base cartMNetwork code to support extra 1k banks and 256b banks; especially the 1k banks. The code changes rippled through to require updates the other 8k and 16k variants too and their Debug widgets. I had basically changed every M Network source code file to do this change.
- While it is easy to add new ROM without these changes, I was more interested in RAM expansion.
- Since this request, I have since discovered and moved onto using 3E+ bankswitching for my new development.

Apologies for so much verbosity.

Regards,
Marco J
